File: packages/ts/src/infra.ts
/** Infrastructure marker detection: CI, containers, orchestration. */
import type { Infrastructure } from "./types.js";
import type { SignalMatrix } from "./matrix.js";
import { fileMatches } from "./signals.js";

export function resolveInfrastructure(matrix: SignalMatrix, files: string[]): Infrastructure {
  const ci = new Set<string>();
  const containers = new Set<string>();
  const orchestration = new Set<string>();
  for (const inf of matrix.infrastructure) {
    // `.github/workflows` is a directory marker: match any file beneath it.
    const dirMarker = !inf.glob.startsWith("*.") && inf.glob.includes("/") && !inf.glob.includes(".");
    const hit = dirMarker
      ? files.some((f) => f === inf.glob || f.startsWith(inf.glob + "/"))
      : files.some((f) => fileMatches(f, inf.glob));
    if (!hit) continue;
    if (inf.category === "ci") ci.add(inf.tool);
    else if (inf.category === "containers") containers.add(inf.tool);
    else orchestration.add(inf.tool);
  }
  return {
    ci: [...ci].sort(),
    containers: [...containers].sort(),
    orchestration: [...orchestration].sort(),
  };
}

File: packages/py/src/repo_fingerprint/infra.py
"""Infrastructure marker detection: CI, containers, orchestration."""
from __future__ import annotations

from typing import Any

from .signals import file_matches
from .types import Infrastructure


def resolve_infrastructure(matrix: dict[str, Any], files: list[str]) -> Infrastructure:
    ci: set[str] = set()
    containers: set[str] = set()
    orchestration: set[str] = set()
    for inf in matrix["infrastructure"]:
        glob = inf["glob"]
        dir_marker = not glob.startswith("*.") and "/" in glob and "." not in glob
        if dir_marker:
            hit = any(f == glob or f.startswith(glob + "/") for f in files)
        else:
            hit = any(file_matches(f, glob) for f in files)
        if not hit:
            continue
        if inf["category"] == "ci":
            ci.add(inf["tool"])
        elif inf["category"] == "containers":
            containers.add(inf["tool"])
        else:
            orchestration.add(inf["tool"])
    return Infrastructure(
        ci=sorted(ci), containers=sorted(containers), orchestration=sorted(orchestration)
    )
